### Basit Tekli Doğrusal Regresyon
- Formül: ŷ = b + w × x
- ŷ = tahmin edilen değer
- x = bağımsız değişken (örneğin Tv Harcaması)
- w = katsayı/eğim  (x (bağımsız değişkendeki) bir birim artışın y'deki (bağımlı değişkene) yansıması)
- b = bias/intercept  (x sıfır olduğunda y'nin başlangıç değeri)

Bir lineer regresyon modeli OLS (Ordinary Least Squares) -- En Küçük Kareler Yöntemini kullanarak hataların karelerinin toplamını minimize eden doğrusal çizgiyi bulmaya çalışır.
- Hata = Σ(y_gerçek - ŷ_tahmin)²

Bir Doğrusal Regresyon modelinin doğru sonuçlar verebilmesi adına veri setinin sağlaması gereken bazı koşullar vardır.Bunlar aşağıda verilmiştir.

### Lineer Regresyonun 5 TEMEL VARSAYIMI
- 1) Doğrusallık (Linearity)
Bağımsız değişken (x) ile bağımlı değişken (y) arasındaki ilişki doğrusal olmalı. Eğer ilişki eğrisel ise (örneğin U şeklinde), düz bir çizgi bu ilişkiyi yanlış temsil eder.
- 2) Bağımsızlık (Independence):
Gözlemler birbirinden bağımsız olmalı, bir gözlemin değeri, diğerini etkilememeli. (Zaman serisi verilerinde bu genelde ihlal edilir, çünkü bugünkü değer dünkü değerden etkilenir.)
- 3) Homoscedasticity (Sabit Varyans)
Hataların (residuals) varyansı, x'in tüm değerlerinde sabit olmalı yani model küçük x değerlerinde de büyük x değerlerinde de benzer büyüklükte hata yapmalı. (Tersi duruma "heteroscedasticity" denir.)
- 4) Normallik (Normality of Residuals)
Hatalar (residuals), yaklaşık olarak normal dağılım göstermeli. Bu, güven aralıkları ve hipotez testlerinin geçerli olması için gerekli.
- 5) Multicollinearity Olmaması (Çoklu Bağlantı Yokluğu)
Eğer birden fazla bağımsız değişken varsa (Multiple Regression'da), bu değişkenler birbirleriyle çok yüksek korelasyonlu olmamalı eğer değişkenler birbirleriyle çok yüksek düzeyde bir korelasyonel ilişkideyse bu durumda model hangi değişkenin model üzerinde etkisi olup olmadığını saptayamaz çünkü bu değişkenler birbirleriyle birlikte hareket eder.

In [1]:
# Lineer Regresyon için Seaborn kütüphanesinden mpg veri setini kullanacağız. 
# Bağımsız değişken beygir gücü (horsepower) bağımlı değişken ise yakıt verimliliği (mpg) olacak.
import seaborn as sns
df = sns.load_dataset('mpg')
df = df.dropna(subset=['horsepower'])

In [2]:
# Modelimizi kuralım
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
X = df[['horsepower']]
y = df['mpg']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
model = LinearRegression()
model.fit(X_train, y_train) # modelin örüntüleri anlayabilmesi için modeli eğitim verisiyle fit (eğitiyoruz) ediyoruz.
print(f"w (eğim): {model.coef_[0]:.4f}") # modele ait katsayı(lar)
print(f"b (intercept): {model.intercept_:.4f}") # modele ait sabit terim.
# model performansını değerlendirmek için R2 skorunu import ediyoruz.
from sklearn.metrics import r2_score
y_pred = model.predict(X_test)
r2 = r2_score(y_test, y_pred)
print(f"R²: {r2:.4f}")

w (eğim): -0.1626
b (intercept): 40.6061
R²: 0.5660


Yukarıdaki sonuçlara baktığımızda beygir gücü 1 birim arttığında yakıt verimliliği 0.1626 düşüyor. Ayrıyeten hiç beygir gücü kullanmazsak yaklaşık 40.61 yakıt verimliliği elde ediyoruz.Son olarak bağımsız değişken (beygir gücü), bağımlı değişken (yakıt verimliliği) üzerindeki değişimin %56.6'sını açıklayabilmektedir.